### ISSUE TITLE
HTTP calls done through the client fail if url has a trailing slash

### ISSUE THREAD

**Issue by amaltaro** [dmwm/DBSClient#55]
While testing dbs3-client 4.0.2, I noticed that the very basic `serverinfo` call fails in one very specific scenario, which is when the `DbsApi` class is instantiated with an url ending with a `/` slash character. While it works just fine when the url doesn't not have a trailing slash.

Here is how to reproduce this problem:
```
>>> from dbs.apis.dbsClient import DbsApi
>>> dbsApi = DbsApi(url = 'https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader/')
>>> dbsApi.serverinfo()
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.5.4.patch4-comp2/sw.amaltaro/slc7_amd64_gcc630/cms/py3-dbs3-client/4.0.2/lib/python3.8/site-packages/dbs/apis/dbsClient.py", line 474, in __parseForException
    data = json.loads(data)
  File "/data/srv/wmagent/v1.5.4.patch4-comp2/sw.amaltaro/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/data/srv/wmagent/v1.5.4.patch4-comp2/sw.amaltaro/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/data/srv/wmagent/v1.5.4.patch4-comp2/sw.amaltaro/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

while this works just fine:
```
>>> dbsApi = DbsApi(url = 'https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader')
>>> dbsApi.serverinfo()
[{'dbs_version': 'v00.04.11', 'server': 'dbs2go git=v00.04.11 go=go1.17.5 date=2021-11-12'}]
```

I am not sure whether this is a problem with the client itself, or with the redirect rules put in the frontend. 
Anyhow, I'm providing a fix for DBSClient in a few minutes.

**Comment by vkuznet** [comment]
@amaltaro , very interesting observation, but according to curl:
```
curl -L -k --key ~/.globus/userkey.pem --cert ~/.globus/usercert.pem https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader//serverinfo
[{"dbs_version":"v00.04.11","server":"dbs2go git=v00.04.11 go=go1.17.5 date=2021-10-12"}]
```
So, I would say it is something in a Python stack rather in our infrastructure. I strongly suggest to test every URL/HTTP related issue with curl since DBSClient provides a wrapper around other Python tools/libraries which potentially may have some issues. But also if you run curl with verbose output (`-v` flag) you can see actual end-point which it reaches on backend server, e.g. `> GET /dbs/int/global/DBSReader/serverinfo HTTP/1.1`.

**Comment by amaltaro** [comment]
I am not sure, but it looks like curl (as well as browsers) follow the redirect URI, hence the request succeeds.

By the way, I just managed to reproduce the same problem with our current agents using DBSClient `3.17.6-comp`. However, if I switch the URL to the production DBSReader, it works just fine:
```
>>> from dbs.apis.dbsClient import DbsApi
>>> dbsApi = DbsApi(url = 'https://cmsweb.cern.ch/dbs/prod/global/DBSReader/')
>>> dbsApi.serverinfo()
{'dbs_instance': 'prod/global', 'dbs_version': '3.16.0-comp5'}
```

it's definitely a problem inserted with the latest frontend changes routing requests to the new Go backend, even though I couldn't spot any issues on those few `deployment` repo changes. @muhammadimranfarooqi might be able to identify something.


Due to this problem, our testbed setup is broken and a few users are unable to keep their activities.

I still think this PR improves the client and it would be great to have it, but probably after the deployment/frontends investigation is over.
@vkuznet @yuyiguo for your consideration.

**Comment by vkuznet** [comment]
@amaltaro , once again please pay attention to curl reporting:
```
curl -v -k \
     --key ~/.globus/userkey.pem --cert ~/.globus/usercert.pem \
     https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader//serverinfo
...
> GET /dbs/int/global/DBSReader//serverinfo HTTP/1.1
> Host: cmsweb-testbed.cern.ch
> User-Agent: curl/7.59.0
> Accept: */*
>
< HTTP/1.1 301 Moved Permanently
...
< Location: /dbs/int/global/DBSReader/serverinfo
```
which means that server tells client (curl) that asked resources is moved to Location. Now, if you use `-L` flag with curl, it will follow the redirect, e.g.
```
 curl -v -L -k --key ~/.globus/userkey.pem --cert ~/.globus/usercert.pem \
     https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader//serverinfo
...
> GET /dbs/int/global/DBSReader//serverinfo HTTP/1.1
...
< HTTP/1.1 301 Moved Permanently
...
< Location: /dbs/int/global/DBSReader/serverinfo
...
> GET /dbs/int/global/DBSReader/serverinfo HTTP/1.1
...
< HTTP/1.1 200 OK
...
[{"dbs_version":"v00.04.11","server":"dbs2go git=v00.04.11 go=go1.17.5 date=2021-12-12"}]
```

So, the actual problem is that Python client does not follow suggested route. But, indeed Go server introduced the issue since route `//serverinfo` is not equal `/serverinfo` (or any other API like `//datatiers` is not equal to `/datatiers`). These are two different names from Go point of view. What it does it provides a redirect location.

Now, about the proper fix. I doubt that your [PR56](https://github.com/dmwm/DBSClient/pull/56) is proper solution. In python, there are different ways to handle HTTP redirect, e.g. [urllib library](https://docs.python.org/3/library/urllib.request.html) provides `rllib.request.HTTPRedirectHandler` class to handle this.  Pycurl which is wrapper around CURL library itself uses curl options, e.g. `curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true)`. Since DBS Client relies on pycurl I suggest to add proper option over there, instead of your PR. For other clients, they should use proper tool/library to follow the redirect location.

**Comment by amaltaro** [comment]
Thanks for following this up, Valentin.

It's important to mention that there is no "python stack" per se here. The dbsclient encapsulates everything, including the HTTP library in the RestClient repository (aparently using urllib and pycurl).

Regarding the proper fix, yes there are different ways to fix it and the "proper" one isn't very clear to me. However, if we have the option to reach the exact resource without any redirect, to me this seems to be the most performant fix. In addition to that, hard-wiring the pycurl flag to follow the redirects in RestClient might not be the best fix in the long run.

So I would insist that stripping the ending `/` character within the `DbsApi` object instantion will be the best fix we can provide, and the most performant. Am I missing anything? Or would you agree with that?

**Comment by vkuznet** [comment]
Of course, stripping extra `/`, or better say, not constructing such double slash end-point(s) is the best way. It is a client job to use correct URL at the end. And, I think both solutions should be put in place to avoid potential issues in a future.

**Comment by vkuznet** [comment]
I added appropriate PR for PycurlClient, see https://github.com/dmwm/PycurlClient/pull/9 and verified that with this change DBSClient correctly behave, i.e. follow redirect link.

I suggest that @yuyiguo should review both PRs and merge them.

**Comment by amaltaro** [comment]
Thanks Valentin. It looks good to me!

**Comment by amaltaro** [comment]
@yuyiguo Hi Yuyi, once you review and merge these 2 PRs, could you please release these two packages in PyPi and let me know the tags? We will then start using the new tag in cmsweb-testbed, which is not functional at the moment.

**Comment by yuyiguo** [comment]
@amaltaro @vkuznet I am reading all the comments and will get the new release soon. #56 is still needed ? Should the pycurl [pr](https://github.com/dmwm/PycurlClient/pulls/9) fix the "//"?

I have a meeting now. Will be back after that.

**Comment by vkuznet** [comment]
@yuyiguo , the two PRs represent different issues, the PR56 to avoid having `//` in URLs, while pycurl PR to follow redirect link. Even though they fix this particular issue, they are different in nature and may not be related. For instance, if HTTP server introduced a new API and would like to keep old one for backward compatibility then it can issue redirect location request, which has nothing to do with `//` issue in URL itself.

**Comment by yuyiguo** [comment]
@vkuznet , 
There two issues here: 
1. FE redirect,  look into prod DBS
 ```
curl -v -L -k --key ~/.globus/userkey.pem --cert ~/.globus/usercert.pem https://cmsweb.cern.ch/dbs/prod/global/DBSReader//serverinfo

GET /dbs/prod/global/DBSReader//serverinfo HTTP/1.1
* Connection #0 to host cmsweb.cern.ch left intact
{"dbs_instance": "prod/global", "dbs_version": "3.16.0-comp5"}
```
There is no redirection in prod. How the testbed FE redirect different from prod FE? Is the redirecting necessary? 

2. Allow double '/' in url or not. 
`curl -v -L -k --key ~/.globus/userkey.pem --cert ~/.globus/usercert.pem https://cmsweb.cern.ch/dbs//prod/global/DBSReader/serverinfo ` works 
`curl -v -L -k --key ~/.globus/userkey.pem --cert ~/.globus/usercert.pem https://cmsweb.cern-testbed.ch/dbs//int/global/DBSReader/serverinfo` not working
Should be the urls allow to have double `/`? If I go to a random website , say `https://cms-http-group.web.cern.ch/cms-http-group/tutorials/index.html `or `https://cms-http-group.web.cern.ch/cms-http-group//tutorials/index.html`, they both work. 
So I would suggest you upgrade Go server to allow double /.

**Comment by yuyiguo** [comment]
@vkuznet 
I didn't mix thing up. I think you did not understand my comments. Do you have a time for a quick chat? I am on zoom. You have my zoom link, right?

**Comment by yuyiguo** [comment]
Hi @amaltaro ,

@vkuznet and I talked on zoom. The key issue here is the modern technology (python Flask or Go) will not support more `/` than what is defined in the code. However, cherrypy will automatically/silently remove the extra `/`  .

For example, `https://cmsweb.cern.ch/dbs///prod/global//DBSReader//serverinfo` works seemless without additional source code. But it is not the case with the modern framework. The redirection [pr](https://github.com/dmwm/PycurlClient/pull/9). will only redirect if "//" is the last one.   If the "//" is in any other places like the the above url, it will give 4xx error instead of 3xx. In addition,  PR #56 will only fix the last "//" too. 

So the I merged the redirection PR to cover the situation of "//" at the end for making the things to work for now. The real improvement  has to come from that we build future url follow the standard. That means a correct DBS url is `https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader` without the endding `\`.  

I uploaded pycurlClient 3.17.7 and DBSClient 4.0.4 to pypi.  Here are some of the simple tests:
 ```
python3
>>> from dbs.apis.dbsClient import *
>>> url="https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader/"
>>> dbs3api = DbsApi(url=url)
>>> dbs3api.serverinfo()
[{'dbs_version': 'v00.04.11', 'server': 'dbs2go git=v00.04.11 go=go1.17.5 date=2021-10-12'}]
>>> url="https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader"
>>> dbs3api = DbsApi(url=url)
>>> dbs3api.serverinfo()
[{'dbs_version': 'v00.04.11', 'server': 'dbs2go git=v00.04.11 go=go1.17.5 

```
Alan, Let me know if this is Ok for you. You don't to have to change DBS urls for now everywhere.  But when the team is editing a file and happening to see a url that has extra `\`, it would be nice to remove so no redirection needed.

**Comment by amaltaro** [comment]
@yuyiguo again, it's not about having additional slashes scattered in the url. This issue was only created - and it's only an issue - because we have an extra slash at THE END of the url:
```
https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader/
```
instead of
```
https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader
```

> That means a correct DBS url is https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader without the endding \.

exactly, instead of asking every single user/application to update their urls - which have been working for many years now - why can't we simply have this *backwards* compatible change in the dbs3-client and ensure that the proper url is used and constructed when building the API endpoint urls?

As already stated in this long discussion, the use of redirects will actually trigger a second HTTP call to the correct resource, please see:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections

Anyhow, I'm done discussing this issue, I think we have more to lose than to gain by keeping this fix out of DBSClient.
Thanks for all the follow up.

**Comment by yuyiguo** [comment]
@amaltaro I see your point. The current release is working and I will not close #56. We will merge it and rebuild the release next time we have new stuff.  Is this sounds a plan to you?

**Comment by amaltaro** [comment]
@yuyiguo yes, that plans sounds good to me. We have just started testing this new client, so it's quite possible that other issues will pop up.

However, I would recommend against using 4.0.4 for too long in external services like ReqMgr2, WMAgent, Unified, McM, etc. Fix #56 will avoid the url redirect and decrease the DBS load almost by half (if most of the APIs have to follow redirects).
Thanks

**Comment by yuyiguo** [comment]
already fixed and #56 will be merged.
